### PR TITLE
Fix table name usage in clauses

### LINKED ISSUE #55: Where on join query with alias not work
# Where on join query with alias not work
When a where clause apply on join query with alias, it cannot parse the correct where.

Let us consider the following query.  

__PHP Query__
```PHP
$query = Entry::query()
  // join self as e
  ->join('entries as e',fn($join) => $join
  ->whereColumn('e.id', 'entries.id')
  ->where('e.collection', 'properties')
  ->where('e.published', 1)
);
// join locations
$query->leftJoin('entries as locations', function($join) {
    $join
        ->where('locations.collection', 'locations')
        ->on('locations.id', 'e.data->location')
        ;
});
// where on joint clause
$query
  // this is expected to search on the JSON data field for property_type
  ->where('e.data->property_type', 'lodge')
  // this is expected to search on the locations slug for value 'shaldon'
  ->where('locations.slug', 'shaldon')
;
```

__SQL Generated__

The where property_type putting the e.data in the json key field but not the data field
```SQL
select * from `entries`
  inner join `entries` as `e`
    on `e`.`id` = `entries`.`id` and `e`.`collection` = 'properties' and `e`.`published` = '1'
  left join `entries` as `locations`
    on `locations`.`collection` = 'locations' and `locations`.`id` = json_unquote(json_extract(`e`.`data`, '$."location"'))
  where
    json_unquote(json_extract(`data`, '$."e.data"."property_type"')) = 'lodge' // Wrong SQL
    and json_unquote(json_extract(`data`, '$."locations.slug"')) = 'shaldon' // Wrong SQL
```
__SQL Expected__
```SQL
select * from `entries`
  inner join `entries` as `e`
    on `e`.`id` = `entries`.`id` and `e`.`collection` = 'properties' and `e`.`published` = '1'
  left join `entries` as `locations`
    on `locations`.`collection` = 'locations' and `locations`.`id` = json_unquote(json_extract(`e`.`data`, '$."location"')) 
  where
    json_unquote(json_extract(`e`.`data`, '$."property_type"')) = 'lodge' // SQL expected
    and `locations`.`slug` = 'shaldon' // SQL expected
```
## Cause
The problem caused by the column check in `src/Entries/EntryQueryBuilder.php`.

```PHP
protected function column($column)
{
    if ($column == 'origin') {
        $column = 'origin_id';
    }

    if (! in_array($column, self::COLUMNS)) {
        $column = 'data->'.$column;
    } else {
        $column = parent::column($column);
    }

    return $column;
}
```

### PR DESCRIPTION
**Pull request by maxwkf**
Fixes #55

### CODE REVIEW

**Comment by maxwkf** [comment]
By using the function setApplyColumnCheck, it then bypass the checking on column and apply original where clause.
```php
$query = Entry::query()
        ->setApplyColumnCheck(false)
```

**Comment by ryanmitchell** [comment]
Would a better fix not be to change the `!Str::startsWith('data->'` to be `Str::contains('data->` ?

**Comment by maxwkfcc** [comment]
> Str::startsWith('data->'

This line is not introduced by me.

**Comment by maxwkfcc** [comment]
> contains

Do you mean like this?
```php
    protected function column($column)
    {
        if (! is_string($column)) {
            return $column;
        }

        if ($column == 'origin') {
            $column = 'origin_id';
        }

        if (in_array($column, self::COLUMNS)) {
            if (! Str::contains($column, 'data->')) {
                $column = 'data->'.$column;
            }
        }

        return $column;
    }

```

**Comment by ryanmitchell** [comment]
Yep - basically... one small change (maybe a copy/paste issue on your side).


```
    protected function column($column)
    {
        if (! is_string($column)) {
            return $column;
        }

        if ($column == 'origin') {
            $column = 'origin_id';
        }

        if (! in_array($column, self::COLUMNS)) {
            if (! Str::contains($column, 'data->')) {
                $column = 'data->'.$column;
            }
        }

        return $column;
    }
```

does that change fix your issue?

**Comment by maxwkfcc** [comment]
> startsWith

It looks you are right.  Let me check it up.  Thanks

**Comment by maxwkf** [comment]
> Yep - basically... one small change (maybe a copy/paste issue on your side).
> 
> ```
>     protected function column($column)
>     {
>         if (! is_string($column)) {
>             return $column;
>         }
> 
>         if ($column == 'origin') {
>             $column = 'origin_id';
>         }
> 
>         if (! in_array($column, self::COLUMNS)) {
>             if (! Str::contains($column, 'data->')) {
>                 $column = 'data->'.$column;
>             }
>         }
> 
>         return $column;
>     }
> ```
> 
> does that change fix your issue?

This would generate the following SQL in which the location part is incorrect.
```SQL
select * from "entries" inner join "entries" as "e" on "e"."id" = "entries"."id" and "e"."collection" = ? left join "entries" as "locations" on "locations"."collection" = ? and "locations"."id" = json_extract("e"."data", '$."location"') where json_extract("e"."data", '$."title"') like ? and json_extract("data", '$."locations.slug"') = ?
```

**Comment by ryanmitchell** [comment]
What about adding a check for the table alias... something like:

```
    protected function column($column)
    {
        if (! is_string($column)) {
            return $column;
        }
        
        $table = Str::contains($column, '.') ? Str::before($column, '.') : '';
        $column = Str::after($column, '.');

        if ($column == 'origin') {
            $column = 'origin_id';
        }

        if (! in_array($column, self::COLUMNS)) {
            if (! Str::contains($column, 'data->')) {
                $column = 'data->'.$column;
            }
        }

        return ($table ? $table.'.' : '').$column;
    }
```

**Comment by maxwkf** [comment]
> What about adding a check for the table alias... something like:
> 
> ```
>     protected function column($column)
>     {
>         if (! is_string($column)) {
>             return $column;
>         }
>         
>         $table = Str::contains($column, '.') ? Str::before($column, '.') : '';
>         $column = Str::after($column, '.');
> 
>         if ($column == 'origin') {
>             $column = 'origin_id';
>         }
> 
>         if (! in_array($column, self::COLUMNS)) {
>             if (! Str::contains($column, 'data->')) {
>                 $column = 'data->'.$column;
>             }
>         }
> 
>         return ($table ? $table.'.' : '').$column;
>     }
> ```

Thanks a lot.  You logic is right.  Do you think this is simplier?
```php
    protected function column($column)
    {
        if (! is_string($column)) {
            return $column;
        }

        if ($column == 'origin') {
            $column = 'origin_id';
        }

        if (! in_array($column, self::COLUMNS)) {
            if (! Str::contains($column, 'data->')) {
                if (! Str::contains($column, '.') ) {
                    $column = 'data->'.$column;
                    
                }
            }
        }

        return $column;
    }
```

**Comment by ryanmitchell** [comment]
Simpler yes but it wouldn’t work the same. Stick with what I did :)

**Comment by maxwkf** [comment]
> Simpler yes but it wouldn’t work the same. Stick with what I did :)
> […](#)
> On 5 Sep 2022, at 14:49, maxwkf ***@***.***> wrote: ﻿ What about adding a check for the table alias... something like: protected function column($column) { if (! is_string($column)) { return $column; } $table = Str::contains($column, '.') ? Str::before($column, '.') : ''; $column = Str::after($column, '.'); if ($column == 'origin') { $column = 'origin_id'; } if (! in_array($column, self::COLUMNS)) { if (! Str::contains($column, 'data->')) { $column = 'data->'.$column; } } return ($table ? $table.'.' : '').$column; } Thanks a lot. You logic is right. Do you think this is simplier? protected function column($column) { if (! is_string($column)) { return $column; } if ($column == 'origin') { $column = 'origin_id'; } if (! in_array($column, self::COLUMNS)) { if (! Str::contains($column, 'data->')) { if (! Str::contains($column, '.') ) { $column = 'data->'.$column; } } } return $column; } — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.

Hi Ryan,

Thanks for your help.  Updated according to our discussion.  Do I need to close the pull request or leave it as open?

**Comment by ryanmitchell** [comment]
Leave it open :)

File: src/Entries/EntryQueryBuilder.php [hunk 1]
```diff
@@ -33,17 +33,20 @@ protected function column($column)
...
+        $table = Str::contains($column, '.') ? Str::before($column, '.') : '';
+        $column = Str::after($column, '.');
+
         if ($column == 'origin') {
             $column = 'origin_id';
         }
 
         if (! in_array($column, self::COLUMNS)) {
-            if (! Str::startsWith($column, 'data->')) {
+            if (! Str::contains($column, 'data->')) {
```
Comment (jasonvarga) [inline src/Entries/EntryQueryBuilder.php:44]: The test still passes if I revert this back to `startsWith`. Is it necessary to change it?

Comment (ryanmitchell) [reply src/Entries/EntryQueryBuilder.php:44]: Yeah should be starts with - it’s a better approach

Comment (jasonvarga) [reply src/Entries/EntryQueryBuilder.php:44]: I agree. It seems like maybe the `contains` check just stuck around accidentally while you figured out the solution to the issue?

Comment (ryanmitchell) [reply src/Entries/EntryQueryBuilder.php:44]: Yeah - contains could potentially return false results too. Please change it :)

**Pull request merged**